Device mode transformation (#892)
rudderlabs/rudderstack-docs: docs/features/transformations/faq.mdx

You cannot update an existing `Version: V0` transformation to the current version.

Instead, you can <Link to="/features/transformations/using-transformations#adding-a-transformation">create a new transformation</Link> based on the <Link to="/features/transformations/using-transformations#transformevent-function">`transformEvent` function</Link>, then reconnect your destinations with this transformation. Finally, delete the existing `Version: V0` transformation.

### What are transformation and library revisions?

Using the <Link to ="/api/transformation-api/">Transformations API</Link>, you can create and store a transformation or library as a draft without publishing them. These are called revisions.

Once the revisions meet the test criteria, you can restore and publish them for production use.

## Using Transformations in device mode 

### While sending events in device mode, how should I choose between event filtering and transformations?

You can choose any based on your requirements:

- **<Link to="/sources/event-streams/sdks/event-filtering/">Event filtering</Link>**: When you need to specify which events should be discarded or allowed to flow through to the destination.

> - **Transformations**: When you need to mask the data, enrich events, implement specific actions on events or remove sensitive PII information from the events.

Reviewer: This option also enables filtering events.

We should rather be highlighting the fact that the event filtering based on "event names" takes place within the SDK while transformation is a more generic approach to manipulating the events, which involves a call to an external RudderStack service.

Author: @saikumarrs Confirming my understanding on the same:

Apart from setting 'event filtering' in the dashboard, users can write a transformation to filter events.

So, the actual question is which option to use to filter events: event filtering in the dashboard or writing a transformation:

* In dashboard event filtering, filtering happens within the SDK
* In transformation, a call to an external RudderStack service is made

Reviewer: Dashboard event filtering is purely by "event names". So, it's only applicable for track events. This feature is the best option if users have just this specific use case.

However, transformations is a powerful feature but has it's caveats which we've discussed already in this doc.

> In case you plan to use both the above-mentioned features, your events will be filtered (based on the <Link to="/sources/event-streams/sdks/rudderstack-javascript-sdk/load-js-sdk/#integrationopts">`integrationOpts`</Link> passed along with the event and/or <Link to="/sources/event-streams/sdks/event-filtering/">event filtering</Link> in the RudderStack dashboard) and transformed (based on the applied transformation) before being sent to the destination.

Reviewer: The integrationsOpts object is not relevant here.

> ### Is there any added cost to use the transformations in device mode?

Reviewer: ```suggestion
### Is there any added performance cost to using the transformations in device mode?
```

> Yes, RudderStack sends a network call for each event when the transformations are used in the device mode. It leads to extra overhead and latency.

Reviewer: ```suggestion
Yes, RudderStack makes a network call for each event when the transformations are used in the device mode.
```

Reviewer: The redacted part is implicit.

Reviewer: Also, "for each event" is applicable only for JS SDK. In the case of mobile SDKs, the events are batched.
CC @desusai7 @itsdebs

Author: @desusai7 can you please confirm?

Reviewer 2: Yes, @saikumarrs is right, we do send events in batches to the transformations service.

    
### Can I switch back to the previous version of an SDK while using a transformation in device mode?

- **For Mobile SDKs**: Yes, you can switch back to the previous SDK version. However, the transformation might not be supported in the previous version.
- **For JavaScript SDK**: It depends on the SDK installation method:
    - If installed through RudderStack CDN, it's not possible to downgrade the SDK version.

>     - If installed through NPM, it's possible to downgrade the SDK version and choose whether to use the transformation or not. 

Reviewer: ```suggestion
    - If installed through NPM, it's possible to downgrade the SDK version. However, the transformation might not be supported in the previous version.
```

### Will my event be lost in case of a network error while using a transformation in device mode?

**No events are lost in case of mobile SDK's** as they are saved locally in the database before being sent to the destination. Once sent successfully, they're removed from the database.

> However, there is a chance of events getting lost while sending them via JavaScript SDK in case of a network error.

Reviewer: ```suggestion
However, there is a chance of events getting lost while sending them via JavaScript SDK in case of a network error as they are not persisted. However, the SDK retries multiple times before dropping the event.
```
